cpp/src/explainer/tree_shap.cu

```diff
@@ -0,0 +1,326 @@
+    = ExtractedPath::Create<ThresholdType>();
+  ExtractedPathImpl<ThresholdType>* paths
+    = dynamic_cast<ExtractedPathImpl<ThresholdType>*>(path_container.get());
+
+  std::size_t path_idx = 0;
+  int tree_idx = 0;
+  int num_groups = 1;
+  if (model.task_type == tl::TaskType::kMultiClfGrovePerClass
+      && model.task_param.num_class > 1) {
+    num_groups = model.task_param.num_class;
+  }
+  for (const tl::Tree<ThresholdType, LeafType>& tree : model.trees) {
+    std::vector<int> parent_id(tree.num_nodes, -1);
+    // Compute parent ID of each node
+    for (int i = 0; i < tree.num_nodes; i++) {
+      if (!tree.IsLeaf(i)) {
+        parent_id[tree.LeftChild(i)] = i;
+        parent_id[tree.RightChild(i)] = i;
+      }
+    }
+
+    // Find leaf nodes
+    // Work backwards from leaf to root, order does not matter
+    // It's also possible to work from root to leaf
+    for (int i = 0; i < tree.num_nodes; i++) {
+      if (tree.IsLeaf(i)) {
+        float v = static_cast<float>(tree.LeafValue(i));
+        int child_idx = i;
+        int parent_idx = parent_id[child_idx];
+        const auto inf = std::numeric_limits<ThresholdType>::infinity();
+        while (parent_idx != -1) {
+          double zero_fraction = 1.0;
+          bool has_count_info = false;
+          if (tree.HasSumHess(parent_idx) && tree.HasSumHess(child_idx)) {
+            zero_fraction = static_cast<double>(
+                tree.SumHess(child_idx) / tree.SumHess(parent_idx));
+            has_count_info = true;
+          }
+          if (tree.HasDataCount(parent_idx) && tree.HasDataCount(child_idx)) {
+            zero_fraction = static_cast<double>(
+                tree.DataCount(child_idx)) / tree.DataCount(parent_idx);
+            has_count_info = true;
+          }
+          if (!has_count_info) {
+            RAFT_FAIL("Lacking sufficient info");
+          }
+          // Encode the range of feature values that flow down this path
+          bool is_left_path = tree.LeftChild(parent_idx) == child_idx;
+          if (tree.SplitType(parent_idx) == tl::SplitFeatureType::kCategorical) {
+            RAFT_FAIL("For now only trees with numerical splits are supported. "
+                      "Trees with categorical splits are not supported yet.");
+          }
+          ThresholdType lower_bound =
+            is_left_path ? -inf : tree.Threshold(parent_idx);
+          ThresholdType upper_bound =
+            is_left_path ? tree.Threshold(parent_idx) : inf;
+          int group_id = tree_idx % num_groups;
+          paths->paths.push_back(
+              gpu_treeshap::PathElement<MySplitCondition<ThresholdType>>{
+                path_idx,
+                tree.SplitIndex(parent_idx),
+                group_id,
+                {lower_bound, upper_bound},
+                zero_fraction,
+                v});
+          child_idx = parent_idx;
+          parent_idx = parent_id[child_idx];
+        }
+        // Root node has feature -1
+        {
```

Comment:
Are we setting up a new block just to re-use `group_id`? If so, would it make sense to just move the declaration to the enclosing scope for both blocks?

Reply:
I kept getting a (false) warning about `group_id` not being used, and the current build setting is treating all compiler warnings as errors.